src/openssl.c

```diff
@@ -3249,17 +3249,49 @@ static BIO *getbio(lua_State *L) {
 	bio = lua_touserdata(L, -1);
 	lua_pop(L, 1);
 
-	BIO_reset(*bio);
-
 	return *bio;
+} /* getbioinstance() */
+
+
+static BIO *getbio(lua_State *L) {
+	BIO *bio = getbioinstance(L);
+	BIO_reset(bio);
+	return bio;
 } /* getbio() */
 
 
+static void pushbiostring(lua_State *L) {
+	BIO *bio = getbioinstance(L);
+	char *str;
+	long len = BIO_get_mem_data(bio, &str);
+	lua_pushlstring(L, str, len);
+	BIO_reset(bio);
+} /* pushbiostring() */
+
+
+static int pem_pw_cb(char *buf, int size, int rwflag, void *u) {
+	lua_State *L = (lua_State *) u;
+
+	if (lua_isnil(L, -1))
+		return 0;
+
+	if (lua_isfunction(L, -1))
+		lua_call(L, 0, 1);
```

Comment:
This isn't safe: if the function throws then you corrupt internal openssl state.

Reply:
I think you may want to only support passing in the password as a string directly: passing in a function is fraught with dangers.

Reply:
This is the way the OpenSSL API has been implemented. The callback function makes sense because then you do not need a separate function for checking if the PEM key is encrypted or not.

Reply:
Yes OpenSSL works that way; but as I mentioned: translating that to lua is fraught with danger.
To support a callback here you'll need to call it inside of a `lua_pcall` so that it is safe if the user supplied callback throws.